### DISCUSSION TITLE
Controlling document actions for a specific document

### DISCUSSION THREAD

**Discussion by chsko** [sanity-io/sanity#3341] [Sanity Studio Support] [answered]
In Studio V2 you can set `_experimental_actions` to define what document actions should be available for a specific document. What is the V3 equivalent?

**Comment by p10e** [comment, marked as answer]
Hi @chsko, thanks for checking out v3 of the Studio! We generally recommend using Document Actions instead of `__experimental_actions` in both v2 (see also the Gotcha on [this page](https://www.sanity.io/docs/ui-affordances-for-actions)) and v3. Here are documentation links for both versions:

- V2: https://www.sanity.io/docs/document-actions
- V3: https://beta.sanity.io/docs/platform/studio/v2-to-v3/document-actions-badges

Depending on your use case, you may also want to hide relevant document types in the "New Document"-structure, for example when using [singleton documents](https://www.sanity.io/docs/create-a-link-to-a-single-edit-page-in-your-main-document-type-list). Here's another round of documentation links:

- V2: https://www.sanity.io/docs/initial-value-templates#56b4073ca73a
- V3: https://beta.sanity.io/docs/platform/studio/v2-to-v3/new-document-structure

Please let us know if anything is unclear or if you run into any issues!

**Reply by plsrd** [reply]
Here's a snippet I used to make it so that my singletons only had access to the publish action: 
```
document: {
  actions: (prev, { schemaType }) => {
        if (singleEdits.includes(schemaType)) {
          return prev.filter(prevAction => prevAction.action == 'publish');
        }
        return prev;
      },
}
```
Note that my `singleEdits` are an array that looks like this: 
```export const singleEdits = ['allInputExamples', 'siteSettings', 'sandbox', 'supportedLangs']```

**Reply by chsko** [reply]
Thanks for the replies! @rlpennell's answer was what I was looking for, specifically, so many thanks for the snippet. @p10e, is @rlpennell's answer the intended way of doing this, or are there other ways as well?

**Reply by p10e** [reply]
@rlpennell's snippet is spot on and the recommended way of doing this! We'll make sure to add it to the documentation.

**Reply by chsko** [reply]
Didn't realise until now that you work at Sanity @rlpennell, my bad 🙈

**Reply by alexbchr** [reply]
Just adding to this question, with the new V3, I don't see any way to filter actions for a specific document. The context that is passed to `document.actions` does contain the `schemaType` and `documentId`, which in theory should let me fetch the corresponding document data. However, from the TS types, I feel like this function cannot really return a Promise.

A good solution would be to:
- Be able to return a Promise for the actions (although since every plugin can have some logic here it may result in long processing time).
- OR provide the current document in the context (I guess this would need to be recalled on every change on the document though) :/ 

Just to provide a bit more context on our use case...

We have a `page` document type for webpages in a website. Every page pretty much has the same structure, but we have a `type` field on each page that defines what is the page type (e.g. Regular Page, Home Page). In v2, using Document actions, we were easily able to prevent editors to delete the Home Page, but in V3, I can easily disable the Delete Action on the document type but not for a specific document...

**Reply by quirm** [reply]
This is a part of my config:
```
document: {
  actions: (prev) => {
    console.log(prev);
    return prev;
  },
  productionUrl,
},
```

On rc3 version `prev` returns an empty array for any document.

**Reply by dwkns** [reply]
@quirm There is an[ open issue](https://github.com/sanity-io/sanity/discussions/3796) for this exact problem. No word yet if it's a bug or they have changed the way it works.

**Reply by alexanderkalachev** [reply]
Still no clear explanation how to do simple things in V3, that worked perfectly with __experimental_actions:
– How to prevent creation of new singleton documents through "New document" button in a top bar?
– How to enable publication for changes in singleton, but preventing it for possible new documents of same type?

**Reply by quirm** [reply]
@alexanderkalachev this is what works for me:
```
export default defineConfig({
  ...,
  document: {
    actions: (prev, { schemaType }) => {
      // Allow only `delete`.
      if (schemaType === 'feedback') {
        return prev.filter((obj) => obj.action === 'delete');
      }

      // Allow only `publish` for page and settings documents (singletons).
      if (schemaType === 'page' || schemaType === 'settings') {
        return prev.filter((obj) => obj.action === 'publish');
      }

      // Allow everything for the rest of documents
      return prev;
    },
    newDocumentOptions: (prev /*, { creationContext }*/) => {
      // Remove singletons and feedback from "create new document"
      return prev.filter(
        (templateItem) => !['page', 'feedback', 'settings'].includes(templateItem.templateId)
      );
    },
  },
  ...
};
```

**Reply by alexanderkalachev** [reply]
@quirm Thank you! I figured out it.
